File: client/Assets/Scripts/Schema/CSVData/CSVData.cs
//////////////////////////////////////////////////////////////////////////
//
// 文件：Assets\Scripts\Schema\CSVData\CSVData.cs
// 作者：Xoen Xie
// 时间：2023/06/19
// 描述：CSV数据基类
// 说明：
//
//////////////////////////////////////////////////////////////////////////
using XSF;

namespace XsfScp
{
    public enum CSVDataType
    {
        None = 0,
        Int,        // 有符号32位整形，配置为空是为0
        Uint,       // 无符号32位整形，配置为空是为0
        Ulong,      // 无符号64位整形，配置为空是为0
        Float,      // 浮点，配置为空是为0
        String,     // 字符串，配置为空是为空串
        Bool,       // 布尔值，配置为空为false， 1或者true时为true
        Array,      // 整形数组类型，配置为空无数据 num1:num2:num3
        IDAndCount, // ID和count组合，id1:count1|id2:count2|id3:count3
        Max,
    }

    public class CSVData : ICSVData
    {
        public virtual CSVDataType DataType { get; }

        public virtual void Read(int nRow, int nCol, string sData)
        {

        }

        private static CSVData[] m_Datas;
        public static CSVData GetData(int nType)
        {
            if (m_Datas == null)
            {
                m_Datas = new CSVData[(int)CSVDataType.Max];
            }

            if (m_Datas[nType] == null)
            {
                switch ((CSVDataType)nType)
                {
                    case CSVDataType.Int: m_Datas[nType] = new CSVData_Int(); break;
                    case CSVDataType.Uint: m_Datas[nType] = new CSVData_Uint(); break;
                    case CSVDataType.Ulong: m_Datas[nType] = new CSVData_Ulong(); break;
                    case CSVDataType.Bool: m_Datas[nType] = new CSVData_Bool(); break;
                    case CSVDataType.Float: m_Datas[nType] = new CSVData_Float(); break;
                    case CSVDataType.Array: m_Datas[nType] = new CSVData_Array(); break;
                    case CSVDataType.String: m_Datas[nType] = new CSVData_String(); break;
                    case CSVDataType.IDAndCount: m_Datas[nType] = new CSVData_IDAndCount(); break;

                    default: throw new XSFSchemaLoadException($"CSVData.GetData type error, type={nType}");
                }
            }

            return m_Datas[nType];
        }
    }
}

File: server-cpp/source/schema/src/csv_data/CSVData.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：source/xsf/src/schema/csv_data/CSVData.cpp
// 作者：Xoen Xie
// 时间：2023/08/14
// 描述：CSV数据定义
// 说明：
//
//////////////////////////////////////////////////////////////////////////

#include "CSVData.h"
#include "CSVDataInc.h"
#include "XSF.h"
using namespace xsf;

namespace xsf_scp
{
    CSVData *CSVData::m_Datas[CSVDataType_Max] = { nullptr };

    CSVData * CSVData::GetData(uint8 nType)
    {
        if(m_Datas[nType] == nullptr)
        {
            switch(nType)
            {
            case CSVDataType_Int:  m_Datas[nType] = new CSVData_Int();  break;
            case CSVDataType_Uint:  m_Datas[nType] = new CSVData_Uint();  break;
            case CSVDataType_Ulong:  m_Datas[nType] = new CSVData_Ulong();  break;
            case CSVDataType_Bool:  m_Datas[nType] = new CSVData_Bool();  break;
            case CSVDataType_Float:  m_Datas[nType] = new CSVData_Float();  break;
            case CSVDataType_Array:  m_Datas[nType] = new CSVData_Array();  break;
            case CSVDataType_String:  m_Datas[nType] = new CSVData_String();  break;
            case CSVDataType_IDAndCount:  m_Datas[nType] = new CSVData_IDAndCount();  break;

            default:
                XSF_ERROR("CSVData::GetData type error, type=%d", nType);
                return nullptr;
            }
        }

        return m_Datas[nType];
    }

    CSVData * CSVData::GetDataByName(const char *sName)
    {
        uint8 nType = 0;
        if (strcasecmp(sName, "int") == 0)
            nType = CSVDataType_Int;
        else if (strcasecmp(sName, "uint") == 0)
            nType = CSVDataType_Uint;
        else if (strcasecmp(sName, "ulong") == 0)
            nType = CSVDataType_Ulong;
        else if (strcasecmp(sName, "float") == 0)
            nType = CSVDataType_Float;
        else if (strcasecmp(sName, "bool") == 0)
            nType = CSVDataType_Bool;
        else if (strcasecmp(sName, "string") == 0)
            nType = CSVDataType_String;
        else if (strcasecmp(sName, "array") == 0)
            nType = CSVDataType_Array;
        else if (strcasecmp(sName, "iac") == 0)
            nType = CSVDataType_IDAndCount;
        else
        {
            XSF_ERROR("CSVData GetDataByName name error, name=%s", sName);
        }
        
        return GetData(nType);
    }
}
